spec/core/data_structures.md

```diff
@@ -230,17 +230,20 @@ enum BlockIDFlag {
+The vote extension is not part of the [`CanonicalVote`](#canonicalvote).
+
+| Name               | Type                            | Description                                                                                 | Validation                                                                                           |
+|--------------------|---------------------------------|---------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
+| Type               | [SignedMsgType](#signedmsgtype) | Either prevote or precommit. [SignedMsgType](#signedmsgtype)                                | A Vote is valid if its corresponding fields are included in the enum [signedMsgType](#signedmsgtype) |
+| Height             | uint64                          | Height for which this vote was created for.                                                 | Must be > 0                                                                                          |
+| Round              | int32                           | Round that the commit corresponds to.                                                       | Must be > 0                                                                                          |
+| BlockID            | [BlockID](#blockid)             | The blockID of the corresponding block.                                                     | [BlockID](#blockid)                                                                                  |
+| Timestamp          | [Time](#Time)                   | Timestamp represents the time at which a validator signed.                                  | [Time](#time)                                                                                        |
+| ValidatorAddress   | slice of bytes (`[]byte`)       | Address of the validator                                                                    | Length must be equal to 20                                                                           |
+| ValidatorIndex     | int32                           | Index at a specific block height that corresponds to the Index of the validator in the set. | must be > 0                                                                                          |
+| Signature          | slice of bytes (`[]byte`)       | Signature by the validator if they participated in consensus for the associated bock.       | Length of signature must be > 0 and < 64                                                             |
+| Extension          | slice of bytes (`[]byte`)       | The vote extension provided by the Application. Only valid for precommit messages.          | Length must be 0 if Type != `SIGNED_MSG_TYPE_PRECOMMIT`                                              |
```

Comment:
If it's only valid for precommits, we might also make a rule that it must be unset/empty on prevotes.

Reply:
Yes, that's what I was trying to do in the `Validation` column of the table:

> Length must be 0 if Type != `SIGNED_MSG_TYPE_PRECOMMIT`

But maybe it could have a better wording?